Use DejaVu fonts as default for text and mathtext

zblz opened the pull request:
In preparation for #4636 and #5128, this PR adds options to use DejaVu fonts as mathtext fonts. In case glyphs are missing, the fallback is STIX, given that they will have a more similar weight than CM. 

DejaVu fonts are not included here, they should be added in a PR that sets DejaVu as default replacing Vera (#4636 or replacement). Do not merge before #4636!

edit: this initial message is outdated, the PR does add the DejaVu fonts and mathtext DejaVu options and sets them as default. See below messages for details.

hetland commented:
@mdboom asked to do a PR to add DejaVu, but did not have a chance to do that yet. I believe this takes care of that, so I won't duplicate. Also, I agree that STIX should be the fallback for missing fonts, when DejaVu does not have them. There is also a LaTeX math font called arev that I use for presentations (using LaTeXit) that looks nice (http://mirrors.rit.edu/CTAN/info/Free_Math_Font_Survey/en/survey.pdf). However, I am not sure how complicated it would be to use in matplotlib. But that may be yet another option that could be used instead of falling back to computer modern.

mdboom commented:
Looks good, thanks!

zblz commented:
@hetland: No, this PR does not actually add DejaVu or make it default, it only allows to use `mathtext.fontset = dejavusans` in the configuration (or do `plt.rcParams['mathtext.fontset']='dejavusans'`) and DejaVu Sans will be used in the math sections if `usetex = False`. Do you want me to do it or will you modify #4636? Note that with this PR the only changes you would need in the `mathtext` section of `rcsetup.py` is to set the default of `mathtext.fontset` to `dejavusans`.

AFAIK, the `arev` latex package is just a packaging of Arev (with other glyphs from mathdesign), so should be equivalent to using DejaVu.

hetland commented:
I think the idea was to make DejaVu (was Arev) the default font for text
and math. It would be easier if you could modify your pull request to make
this small change, so it is all in one place.

I agree that the area package is identical to DejaVu (actually, DejaVu may
have more by now), but it may be useful to have latex render fonts as pdf
snippets using arev as a fallback. I'm not sure how may people still use
this option -- I think mathtext should definitely remain the preferred
default rendering engine.

On Thu, Oct 8, 2015 at 3:16 PM, Victor Zabalza <notifications@github.com>
wrote:

> @hetland <https://github.com/hetland>: No, this PR does not actually add
> DejaVu or make it default, it only allows to use mathtext.fontset =
> dejavusans in the configuration (or do
> plt.rcParams['mathtext.fontset']='dejavusans') and DejaVu Sans will be
> used in the math sections if usetex = False. Do you want me to do it or
> will you modify #4636 <https://github.com/matplotlib/matplotlib/pull/4636>?
> Note that with this PR the only changes you would need in the mathtext
> section of rcsetup.py is to set the default of mathtext.fontset to
> dejavusans.
>
> AFAIK, the arev latex package is just a packaging of Arev (with other
> glyphs from mathdesign), so should be equivalent to using DejaVu.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/5214#issuecomment-146674038>
> .
>



-- 
Prof. Rob Hetland
Texas A&M Univ.  –  Dept. of Oceanography
http://pong.tamu.edu/~rob

zblz commented:
Ok, I'll close your PR and change the title for this one.

I'll take a look at consistency with selecting the font for usetex, but if
arev is not in the core TeX distribution, it will mean trouble when users
find LaTeX errors because they don't have arev installed. Right now
sans-serif TeX output defaults to computer modern sans serif (which is
obviously in all TeX distributions), and the only alternative sans-serif
fonts in core TeX are Helvetica and Avant Garde, and I'm not sure about
their math coverage.

I'm not sure how may people still use
> this option -- I think mathtext should definitely remain the preferred
> default rendering engine.
>

Mathtext is definitely preferred --it's much faster and portable-- but
`usetex` has a clear use case for complex math and the typography of
mathtext will never be at the level of TeX. I often use it when I want a
final plot to have the exact same fonts as the paper it will be embedded in.

jkseppan commented:
Just a comment about TeX fonts: I think http://www.tug.dk/FontCatalogue/mathfonts.html is the most complete list of fonts that have TeX math support.

mdboom commented on lib/matplotlib/mpl-data/fonts/ttf/local.conf:
```diff
@@ -1,32 +0,0 @@
-                        <family>Bitstream Vera Sans Mono</family>
-                </prefer>
-        </alias>
-</fontconfig>
```
I actually didn't realize this was here.  I think this should be updated to use DejaVu rather than just deleted.  It sets the default font types if a user is opting for the font-config mode rather than using matplotlib's own font lookup system.

It's not strictly necessary, but removing this will certainly change behavior for those that are relying on it.

zblz replied:
I thought it was part of the Vera release and not being actually used. If it is used I will definitely put it back and modify it.

jenshnielsen replied:
Is there any particular reason for removing the old fonts is stead of just adding the new ones and changing the defaults. For all other changes we have made them backwards compatible and added the old default to the classic style

zblz replied:
DejaVu are a superset of Bitstream Vera, so it would be redundant to ship them both. The classic style can still have `mathtext.fontset : cm`. However, it is true that it might cause problems with matplotlibrc that have been modified to have a single `font.sans-serif : Bitstream Vera Sans` rather than a list...

zblz replied:
Regarding shipping them both, I just realised that even though the DejaVu fonts are 5.3 MB, the Vera fonts are only 600 KB, no from a distribution size perspective it is not a huge difference shipping them both rather than removing Vera.

mdboom replied:
@jenshnielsen: As @zblz said, since DejaVu is a superset of Vera, using DejaVu whenever Vera is requested should have *no* visible impact other than missing characters now being present, which I'd call a bugfix, not a behavior change.  I don't think there's any benefit to keeping the old font, particularly with font subsetting working so well in all backends.

That said, it might be worth adding a built-in mapping from Vera to DejaVu, so that if a user was explicitly setting Vera (rather than just leaving the defaults), and they don't have it installed system wide, they will get DejaVu and things continue to work.  This is the kind of thing where some extra release candidate testing may be necessary to catch all of the environmental gotchas about this change.

mdboom commented on matplotlibrc.template:
```diff
@@ -221,8 +221,9 @@ backend      : %(backend)s
 #mathtext.sf  : sans
-#mathtext.fontset : cm # Should be 'cm' (Computer Modern), 'stix',
-                       # 'stixsans' or 'custom'
+#mathtext.fontset : dejavusans # Should be 'cm' (Computer Modern),
```
Put dejavusans first in the list and say "'dejavusans' (default),"

mdboom commented on lib/matplotlib/tests/test_mathtext.py:
```diff
@@ -177,15 +177,19 @@ def single_test():
 make_set('mathfont', 'stix', font_tests, ['png'])
 make_set('mathfont', 'stixsans', font_tests, ['png'])
+make_set('mathfont', 'dejavusans', font_tests, ['png'])
+make_set('mathfont', 'dejavuserif', font_tests, ['png'])
```
This will need all of the new images added, of course.

mdboom commented:
Note to distributors: @sandrotosi, @jspaleta, others:  I think Debian and Fedora are removing the fonts vendored with matplotlib and requiring the system packages of those fonts instead.  Just a heads up that we are changing our main vendored font from "Bitstream Vera" to "DejaVu" and that package requirement will need to be updated as well.  (And this is expected to land in matplotlib 2.0)

@zblz: This should get a "What's New" entry.

toddrjen commented:
On Oct 13, 2015 2:36 PM, "Michael Droettboom" <notifications@github.com>
wrote:
>
> Note to distributors: @sandrotosi, @jspaleta, others: I think Debian and
Fedora are removing the fonts vendored with matplotlib and requiring the
system packages of those fonts instead. Just a heads up that we are
changing our main vendored font from "Bitstream Vera" to "DejaVu" and that
package requirement will need to be updated as well. (And this is expected
to land in matplotlib 2.0)
>
> @zblz: This should get a "What's New" entry.
>

Is this something they are doing by hand?  If so, might this be something
setup.py could support?

sandrotosi commented:
thanks a ton @mdboom for the heads up! we are indeed (trying to) use the debian font files where possible instead of distributed ones, so it's important to know. @toddrjen yes, at least in debian we are doing it by hand, as we inspect the shipped fonts, then we need to match with the currenctly available debian fonts, add the dependencies to the binary package and then create the symlinks in the mpl namespace

zblz commented:
I have added the mathtext baseline images, squashed a few commits, rebased with master to fix a doc-building warning, and added a what's new file. 

I have changed a bit the mathtext fontset classes to reuse code and fix the fallback for alternative sized symbols, which was broken (and one of the main reasons to fallback onto STIX).

zblz commented:
The advance of the point glyph with this PR is slightly too large. It doesn't happen on regular text. I'll look a bit deeper and try to find the reason.

Tillsten commented:
To support the classic mpl style Vera should not be removed. Btw wonderful work, i really disliked vera.

zblz commented:
@Tillsten -- DejaVu is a superset of Vera, so that the only difference between the new and old default (for regular text) will be when Vera falls back to CM. That could be argued to be a bug, so I'm not sure we should retain the option of a small character set in the classic mpl style.

mdboom commented:
@Tillsten: Deja Vu is the same font as Vera under a new name, with additional glyphs added and bugs fixed (I think historically it was a fork).  There is no need to retain the old font for the classic style.

zblz commented:
The problem with the dot advance I mentioned above is not related to this PR, it is general to mathtext spacing and is being fixed in #5301, so it should not hold this PR.

zblz commented:
@mdboom: re: the comment you left on #5301 about not storing too many baseline images considering that they might have to be regenerated with the fixed freetype version (#5307), what should we do with this PR. It has a lot of new images to be added, but few that are modified, so maybe we should merge this PR after #5307 and use the local freetype to generate the `dejavusans` and `dejavuserif` new mathtext images. Otherwise we have to either a) not create the new baseline images until #5307, and have master fail tests because baseline images are not there, b) as a) but mark them knownfail, c) add the tests for the new images in #5307. I don't like any of these particularly, so probably better to wait until I can rebase this PR on #5307 and generate new images.

mdboom commented:
Yes, that makes sense.

mdboom commented:
@zblz: I think we're in a position to merge this now.  I think what needs to happen is:

1. Rebase this on master to get the default freetype version changes
2. Build your local copy with the special version of freetype
3. Regenerate the new images here (`*dejavu*`) using that version of freetype
4. Amend the last commit here (`add mathtext dejavu baseline images`) with those new images

If you're too busy to do that, I can probably do in the next few days.

Thanks for this!  I think this will really help those with non-Latin-character languages.

jenshnielsen commented:
@mdboom  shouldn't #5306 go in first?

mdboom commented:
@jenshnielsen wrote:
> @mdboom shouldn't #5306 go in first?

@jenshnielsen: Doh.  You're right.  Somehow I thought it had and didn't check.

jenshnielsen commented:
I think it is ready but needs a rebase?

zblz commented:
Yes, I was waiting to rebase on #5306 and regenerate the images.

jenshnielsen commented:
#5306 is in so this can be rebased now

zblz commented:
I rebased and regenerated images, but just realized that there is a problem with some characters in dejavuserif mathtext showing a box rather than falling back to STIX serif. I'll look into it tomorrow.

mdboom commented:
Thanks, @zblz.

zblz commented:
I think that the problem was that under FreeType 2.6.1 the TrueType font would return a `.notdef` symbol name and a square glyph when the glyph was not found, rather than the old behaviour of returning a `KeyError`. Therefore, it was not being caught as a missing glyph. Commit 0ef2cd4 should also fix this for all truetype fonts that fall back onto CM.

mdboom commented:
Actually I think the relevant change was #5410, which replaced the cmap dictionary with calls to `get_char_index`.  `get_char_index` returns 0 when the char doesn't exist -- it doesn't throw a `KeyError` like a dictionary would.  I think you can just check for `gind == 0`, don't need the additional call to `get_glyph_name` and don't need the `try/except`.  That should work for all versions of freetype.

zblz commented:
@mdboom - that makes more sense, I had not checked your latest font PRs. I changed it as you suggested and now everything should be fine.

mdboom commented:
Thanks!

tacaswell commented:
This needs to be backported to 2.0.x?

mdboom commented:
Backported to 2.0.x as 4a2a18689794b6f0c6d35c55e4691e36bc47d5f8
